src/db/roomstore.ts

```diff
@@ -0,0 +1,347 @@
+        const cached = this.entriesMatrixIdCache.get(matrixId);
+        if (cached && cached.ts + ENTRY_CACHE_LIMETIME > Date.now()) {
+            return cached.e;
+        }
+        const entries = await this.db.All(
+            "SELECT * FROM room_entries WHERE matrix_id = $id", {id: matrixId},
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            const remoteId = entry.remote_id as string || "";
+            const row = await this.db.Get(
+                "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                {rid: remoteId},
+            );
+            if (!row) { continue; }
+
+            res.push({
+                id: (entry.id as string),
+                matrix: matrixId ? new MatrixStoreRoom(matrixId) : null,
+                // tslint:disable-next-line no-any
+                remote: remoteId ? new RemoteStoreRoom(remoteId, row as any) : null,
+            });
+        }
+        if (res.length > 0) {
+            this.entriesMatrixIdCache.set(matrixId, {e: res, ts: Date.now()});
+        }
+        return res;
+    }
+
+    public async getEntriesByMatrixIds(matrixIds: string[]): Promise<IRoomStoreEntry[]> {
+        const entries = await this.db.All(
+            `SELECT * FROM room_entries WHERE matrix_id IN ('${matrixIds.join("','")}')`,
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            const matrixId = entry.matrix_id as string || "";
+            const remoteId = entry.remote_id as string || "";
+            const row = await this.db.Get(
+                "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                {rid: remoteId},
+            );
+            if (!row) { continue; }
+
+            res.push({
+                id: (entry.id as string),
+                matrix: matrixId ? new MatrixStoreRoom(matrixId) : null,
+                // tslint:disable-next-line no-any
+                remote: remoteId ? new RemoteStoreRoom(remoteId, row as any) : null,
+            });
+        }
+        return res;
+    }
+
+    public async linkRooms(matrixRoom: MatrixStoreRoom, remoteRoom: RemoteStoreRoom) {
+        await this.upsertRoom(remoteRoom);
+
+        const values = {
+            id: uuid(),
+            matrix: matrixRoom.roomId,
+            remote: remoteRoom.roomId,
+        };
+
+        try {
+            await this.db.Run(`INSERT INTO room_entries VALUES ($id, $matrix, $remote)`, values);
+            log.verbose("Created new entry " + values.id);
+        } catch (ex) {
+            log.error("Failed to insert room entry", ex);
+            throw Error("Failed to insert room entry");
+        }
+    }
+
+    public async setMatrixRoom(matrixRoom: MatrixStoreRoom) {
+        // This no-ops, because we don't store anything interesting.
+    }
+
+    public async getEntriesByRemoteRoomData(data: IRemoteRoomDataLazy): Promise<IRoomStoreEntry[]> {
```

Comment:
this whole method feels clunky

Reply:
I don't think it's that bad? We build up a WHERE clause based on provided data and iterate over the result. Suggestions welcome though, since I might be missing something.

Reply:
not really sure, it just... *feels* clunkly and hard to cache if something happens down the road. Probably fine